SSZipArchive/SSZipArchive.m

```diff
@@ -1014,7 +982,12 @@ + (NSString *)_filenameStringWithCString:(const char *)filename
     if (@available(macOS 10.10, iOS 8.0, watchOS 2.0, tvOS 9.0, *)) {
 #endif
         // supported encodings are in [NSString availableStringEncodings]
-        [NSString stringEncodingForData:data encodingOptions:nil convertedString:&strPath usedLossyConversion:nil];
+        BOOL isloss = NO;
+        NSStringEncoding encGB_18030_2000 = CFStringConvertEncodingToNSStringEncoding(kCFStringEncodingGB_18030_2000);
+        NSStringEncoding encShiftJIS = CFStringConvertEncodingToNSStringEncoding(kCFStringEncodingShiftJIS);
+        NSStringEncoding encDOSLatinUS = CFStringConvertEncodingToNSStringEncoding(kCFStringEncodingDOSLatinUS);
+        NSArray * encList = @[@(encGB_18030_2000), @(encShiftJIS), @(encDOSLatinUS)];
+        [NSString stringEncodingForData:data encodingOptions:@{NSStringEncodingDetectionSuggestedEncodingsKey:encList} convertedString:&strPath usedLossyConversion:&isloss];
```

Comment:
The non-utf8 file will process here. The suggestion encoding set includes GB_18030_2000, ShiftJIS, and DOSLatinUS. More encoding format can be appended. And utf8 flag and platform(made_by) can be ignored.

Reply:
It's tricky to define our own set of suggested encodings (although I know, it's done in present code for macOS 10.9 or older).

I would be in favor of changing our API to have the fallback encoding options as a parameter to unzipping (or as a callback). Normally, well-formed archives should be in unicode, and old archived should be dealt with according to client preferences instead of attempting imperfect encoding detection.

A. If done as an encoding option: that parameter will tell if we fallback non-unicode to a particular encoding, or to risky encoding-detection, or to hexa-strings.
B. If done as a callback option: that parameter will be called with the NSData each time it's not-unicode, and that callback will return the desired NSString.

Note that neither of those solutions (present code and above suggestions) is dealing with duplicate filenames: if an archive has two identical filenames in it (unicode or not), the behavior is possibly undefined (error? overwrite? skip? rename?) and that would also ideally require a callback to let the client decide on what to do.

Reply:
Which solution do you prefer? A or B?